### DISCUSSION TITLE
Change project initialization to not require wails binary

### DISCUSSION THREAD

**Discussion by mpldr** [wailsapp/wails#2291] [Ideas]
One of the great features Go has is `go run`. With this it is possible to run wails without having a binary in the path for other projects which are not using it. (muscle memory makes me type `wails build` anyways)

I would like to propose a `--go-run` flag which does the following:
- create a `tools.go` (see below for explanation)
- create a `Makefile` or a `build.sh`

## tools.go

This file has the contents:

```go
//go:build ignore

package main

import _ "github.com/wailsapp/wails/v2/cmd/wails"
```

This pins the wails version and it ensures that wails is not removed from the dependencies when running `go mod tidy`, otherwise there might be dependency issues. It also allows for multiple projects to use different versions (for example, a hobby project might use a more up-to-date version of wails than a work project)

## Makefile/build.sh

If `make` is available, create a Makefile wrapping the most common functions:

```make
GO ?= $(shell which go || echo go)
WAILS := $(GO) run github.com/wailsapp/wails/v2/cmd/wails

build:
	$(WAILS) build

dev:
	$(WAILS) dev
```

If `make` is not found within the path, a simple `build.sh` achieving the same can be generated. The reason for this is simply because nobody wants to type the import path every time they want to run it.

## Updating wails

Wails can now be updated for just this project the same way all other Go packages are updated and there is no chance of a developer using a wrong version.

**Comment by leaanthony** [comment]
Thanks for opening this. Some of this will be addressed in as-of-yet unannounced v3 👍

**Reply by mpldr** [reply]
Feel free to ping me if this advances, as this also has other implications (like a command to generate only the bindings)

**Reply by leaanthony** [reply]
Yeah, I'm literally looking at that now. Static analysis is hard 😅

**Reply by mpldr** [reply]
Right now it's simply not possible to write a makefile for wails, as the build steps are not atomic.

**Reply by mpldr** [reply]
To elaborate: the `wails build` command performs most of the steps internally (`npm build`, binding generation, compiling), some of which are just not part of the build process (`go mod tidy`)

**Reply by mpldr** [reply]
While I do understand the appeal, this is also pretty limiting.

**Reply by leaanthony** [reply]
Correct

**Comment by johnmaguire** [comment]
Currently v3 is also incompatible with this flow. See [this closed (but valid) bug](https://github.com/wailsapp/wails/issues/4038) for details.
